packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```
+<nimble-table>
+    <nimble-table-column-text valueKey="name"></nimble-table-column-text>
+    <nimble-table-column-icon valueKey="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation.
+
+The table API to support this could look like the following:
+
+```TS
+public Table extends FoundationElement {
+    ...
+
+    @observable
+    public slottedColumns: ITableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `ITableColumn`
+
+This interface is what a column web component (i.e. a slotted column) must implement.
+
+```TS
+interface ITableColumn {
+    header?: string; // the string to use in the header for the column
+    headerTemplateFn?: () => ViewTemplate; // the template to use for the header (string if none provided)
+    cellTemplateFn: () => ViewTemplate<ITableCellData>; // the template to use to render the cell content for the column (required)
+    getDataKeys: () => string[]; // the set of keys the table needs to provide the cellTemplateFn with the proper subset of data
+}
+```
+
+The `ITableColumn` interface would be updated to support other features not covered in this HLD such as sorting and grouping.
+
+### `ITableCellData`
+
+In order for the `cellTemplateFn` in the above interface to be able to act on the necessary table data, we should have a type that the template function can work with directly through.
+
+```TS
+type tableData = string | number | boolean | Date | null | undefined;
+
+interface ITableCellData {
+    data: { [key: string]: tableData };
```

Comment:
Would we expect columns to modify the table data or should we start by marking it read-only? Or is it getting a copy of the data rather than a reference to it?

Reply:
It gets a copy.